parser.rst

```diff
@@ -0,0 +1,884 @@
+for a rule). Usually, PEG parsers avoid this exponential time complexity with a
+technique called "packrat parsing" [1]_ which not only loads the entire
+program in memory before parsing it but also allows the parser to backtrack
+arbitrarily. This is made efficient by memoizing the rules already matched for
+each position. The cost of the memoization cache is that the parser will naturally
+use more memory than a simple LL(1) parser, which normally are table-based. We
+will explain later in this document why we consider this cost acceptable.
+
+
+Key ideas
+~~~~~~~~~
+
+.. important::
+    Don't try to reason about a PEG grammar in the same way you would to with an EBNF
+    or context free grammar. PEG is optimized to describe **how** input strings will
+    be parsed, while context-free grammars are optimized to generate strings of the
+    language they describe (in EBNF, to know if given string is in the language you need
+    to do work to find out as is not imediately obvious from the grammar).
+
+* Alternatives are ordered ( ``A | B`` is not the same as ``B | A`` ).
+* If a rule returns a failure doesn't mean that the parsing has failed,
+  it just means "try something else".
+* By default PEG parsers run in exponential time and they are keep linear By
+  using memoization.
+* If parsing fails completely (no rule can suceed parsing all the input text), the
+  PEG parser doesn't have a concept of "where the syntax error is".
+
+
+Consecuences or the ordered choice operator
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+.. _consecuences-of-ordered-choice:
+
+Although PEG may look like EBNF, but its meaning is quite different. The fact
+that in PEG parsers alternatives are ordered (which is the core of how PEG
+parsers work) has deep consecuences, other than removing ambiguity.
+
+If a rule has two alternatives and the first of them suceeds, the second one is **not** attempted
+even if the caller rule failes to parse the rest of the input. This is often described that the parser
+is "eager". To illustrate, this, consider the following two rules: ::
+
+    first_rule:  ( 'a' | 'aa' ) 'a'
+    second_rule: ('aa' | 'a'  ) 'a'
+
+In a relgular EBNF grammar, both rules specify the language ``{aa, aaa}`` but
+in PEG, one of these two rules accepts the string ``aaa`` but not the string
+``aa``. The other does the opposite -- it accepts the string the string ``aa``
+but not the string ``aaa``. The reason is that the rule ``('a'/'aa')'a'`` does
+not accept ``aaa`` because ``'a'/'aa'`` consumes the first ``a``, letting the
+final ``a`` in the rule consume the second, and leaving out the third ``a``.
+As the rule has suceeded, no attempt is ever made to go back and let
+``'a'/'aa'`` try the second alternative. The expression ``('aa'/'a')'a'`` does
+not accept ``aa`` because ``'aa'/'a'`` accepts all of ``aa``, leaving nothing
+for the final ``a``. Again, the second alternative of ``'aa'/'a'`` is not
+tried.
```

Comment:
A subtle point here is that catenation does not distribute across choices.  I presume that ``'aaa'|'aa' *will* match ``aa`` because upon failure of ``aaa`` to match, ``aa`` will be tried, so that the pattern is *not* equal to ``('aa'/'a')'a'``.

Reply:
> so that the pattern is *not* equal to ``('aa'/'a')'a'``.

Indeed, that is the consequence of the parser being "eager" and the ordered alternative operator. The key here is that groping is its own rule (so when you write `('aa'/'a')'a'` you are writing `rule_1 'a'`. And the alternative eagerness is evaluated at the rule granularity